### PR TITLE
WorldMap: Add agility shortcut level requirements

### PR DESCRIPTION
**Pull request by MESLewis**
image

Add shortcut type and level requirement to world map tooltips.
Overlay image on shortcuts that are above your agility level

image Courtesy of @oplosthee 
A huge thank you to @trimbe for their significant help on gathering the shortcut locations!

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapConfig.java [hunk 1]
```diff
@@ -57,4 +57,26 @@ default boolean fairyRingIcon()
...
+		description = "Display the required Agility level in the icon tooltip",
+		position = 3
+	)
+	default boolean agilityShortcutTooltips()
+	{
+		return true;
+	}
+
+	@ConfigItem(
+			keyName = WorldMapPlugin.CONFIG_KEY_AGILITY_SHORTCUT_LEVEL_ICON,
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapConfig.java]: Fix your indentation

File: runelite-client/src/main/java/net/runelite/client/plugins/worldmap/AgilityShortcutLocation.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+	BRIMHAVEN_DUNGEON_EAST_STEPPING_SOUTH("Stepping Stones", 83, new WorldPoint(2693, 9529, 0)),
+	ELVEN_ADVANCED_CLIFF_SCRAMBLE("Rocks", 85, new WorldPoint(2337, 3253, 0)),
+	KALPHITE_WALL("Crevice", 86, new WorldPoint(3214, 9508, 0)),
+	BRIMHAVEN_DUNGEON_VINE_EAST("Vine", 87, new WorldPoint(2672, 9582, 0)),
+	BRIMHAVEN_DUNGEON_VINE_WEST("Vine", 87, new WorldPoint(2606, 9584, 0)),
+	RENEVANT_CAVES("Jump", 89, new WorldPoint(3240, 10144, 0));
+
+	private final String tooltip;
+	private final WorldPoint locaiton;
+	private final AgilityShortcutPoint agilityShortcutPoint;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/worldmap/AgilityShortcutLocation.java]: Move this outside of this, this is not constant.

File: runelite-client/src/main/java/net/runelite/client/plugins/worldmap/AgilityShortcutLocation.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+	AgilityShortcutLocation(String description, int level, WorldPoint location)
+	{
+		this.tooltip = description + " - Level " + level;
+		this.locaiton = location;
+		this.agilityShortcutPoint = new AgilityShortcutPoint(tooltip, location);
+		this.levelReq = level;
+	}
+
+	static void setIcon(AgilityShortcutLocation[] locations, BufferedImage image)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/worldmap/AgilityShortcutLocation.java]: Remove this. You are basically modifying constants.

File: runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java [hunk 4]
```diff
@@ -106,6 +124,28 @@ public void onConfigChanged(ConfigChanged event)
 				case CONFIG_KEY_FAIRY_RING_ICON:
 					FairyRingLocation.setIcon(config.fairyRingIcon() ? FAIRY_TRAVEL_ICON : BLANK_ICON);
 					break;
+				case CONFIG_KEY_AGILITY_SHORTCUT_TOOLTIPS:
+					if (config.agilityShortcutTooltips())
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java]: What I would do here is
```java
case CONFIG_KEY_AGILITY_SHORTCUT_TOOLTIPS:
case CONFIG_KEY_AGILITY_SHORTCUT_LEVEL_ICON:
worldMapPointManager.removeIf(AgilityShortcutPoint.class::isInstance);

if (config.agilityShortcutTooltips())
{
  Arrays.stream(AgilityShortcutLocation.values()).map(value -> new AgilityShortcutPoint(value.getLocation(), config.agilityShortcutLevelIcon() || value.getLevelReq() >= playerLevel ? BLANK_ICON : NOPE_ICON)).forEach(worldMapPointManager::add)
}
```

File: runelite-client/src/main/java/net/runelite/client/plugins/worldmap/AgilityShortcutPoint.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import lombok.Getter;
+import net.runelite.client.ui.overlay.worldmap.WorldMapPoint;
+
+public class AgilityShortcutPoint extends WorldMapPoint
+{
+	@Getter
+	private static final List<AgilityShortcutPoint> points = new ArrayList<>();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/worldmap/AgilityShortcutPoint.java]: This is just moving of the problem, basically.

File: runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java [hunk 6]
```diff
@@ -121,11 +148,35 @@ protected void startUp() throws Exception
...
+		if (event.getSkill() == Skill.AGILITY)
+		{
+			int newAgilityLevel = Experience.getLevelForXp(client.getSkillExperience(Skill.AGILITY));
+			if (config.agilityShortcutLevelIcon() && newAgilityLevel > agilityLevel)
+			{
+				agilityLevel = newAgilityLevel;
+
+				Arrays.stream(AgilityShortcutLocation.values())
+					.map(value -> new AgilityShortcutPoint(value, config.agilityShortcutLevelIcon() && value.getLevelReq() > agilityLevel ? NOPE_ICON : BLANK_ICON))
+					.forEach(worldMapPointManager::add);
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java]: Does this not need to remove the old icon first?

File: runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java [hunk 7]
```diff
@@ -121,11 +148,35 @@ protected void startUp() throws Exception
...
+		worldMapPointManager.removeIf(AgilityShortcutPoint.class::isInstance);
+	}
+
+	@Subscribe
+	public void onXpChanged(ExperienceChanged event)
+	{
+		if (event.getSkill() == Skill.AGILITY)
+		{
+			int newAgilityLevel = Experience.getLevelForXp(client.getSkillExperience(Skill.AGILITY));
+			if (config.agilityShortcutLevelIcon() && newAgilityLevel > agilityLevel)
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java]: People can hop accounts etc so this should probably be !=?

**Comment by MESLewis** [comment]
Merge conflicts have been resolved.

**Pull request merged**